### PR TITLE
Fixed memory leaks in Exception class

### PR DESCRIPTION
**Pull request by sebastianbaginski**
Keep track of stack frames for correct cleanup.
Fixed memory leak in `Stack` class.

### CODE REVIEW

File: source/sdk/lang/Exception.ooc [hunk 1]
```diff
@@ -47,6 +49,18 @@ _pushStackFrame: inline func -> StackFrame {
     buf
 }
 
+_markStackFrame: inline func (frame: StackFrame) {
```
Comment (davidhesselbom) [inline source/sdk/lang/Exception.ooc]: Just so you know, `inline` doesn't actually work.

File: source/sdk/lang/Exception.ooc [hunk 2]
```diff
@@ -34,6 +34,8 @@ exceptionStack := ThreadLocal<Stack<StackFrame>> new()
 _exception := ThreadLocal<Exception> new()
 _EXCEPTION: Int = 1
 
+stackFramesToDelete := Stack<StackFrame> new()
```
Comment (davidhesselbom) [inline source/sdk/lang/Exception.ooc]: I know this file was already a mess, but, does this really need to be a global thing? Could it be put inside a class and made static, perhaps?

File: source/sdk/lang/Exception.ooc [hunk 3]
```diff
@@ -47,6 +49,18 @@ _pushStackFrame: inline func -> StackFrame {
...
 
+_markStackFrame: inline func (frame: StackFrame) {
+  stackFramesToDelete push(frame)
+}
+
+_cleanupStackFrames: inline func {
+  while (stackFramesToDelete empty?() == false) {
+    frame := stackFramesToDelete pop() as StackFrame
+    if (frame)
+      gc_free(frame as Void*)
```
Comment (davidhesselbom) [inline source/sdk/lang/Exception.ooc]: I wonder if it's possible to define a `free` function on the `StackFrame` cover so you could call `frame free()` instead...

File: source/sdk/lang/Exception.ooc [hunk 4]
```diff
@@ -142,6 +155,15 @@ Exception: class {
         init(null, message)
     }
 
+    free: override func {
+      while (this backtraces empty?() == false) {
+        bt := this backtraces pop() as Backtrace
```
Comment (davidhesselbom) [inline source/sdk/lang/Exception.ooc]: `backtrace`, please?

File: source/sdk/lang/Exception.ooc (hunk 1 above)
Comment (sebastianbaginski) [inline source/sdk/lang/Exception.ooc]: Good to know.
I was just following the conventions :)

File: source/sdk/lang/Exception.ooc (hunk 2 above)
Comment (sebastianbaginski) [inline source/sdk/lang/Exception.ooc]: Let's clear this up in another PR / debt issue, along with all other global stuff in Exception / Backtrace. What do you think ?

Comment (davidhesselbom) [inline source/sdk/lang/Exception.ooc]: Fine.

File: source/sdk/lang/Exception.ooc (hunk 4 above)
Comment (sebastianbaginski) [inline source/sdk/lang/Exception.ooc]: Fixed.

**Comment by sebastianbaginski** [comment]
Updated. Now it releases the stack frame in `_popStackFrame`, which is automatically called by the compiler.
`Exception` class uses another function to take frame from the frame stack, so now everything is released even when no exception is thrown.

File: source/sdk/lang/Exception.ooc (hunk 3 above)
Comment (sebastianbaginski) [inline source/sdk/lang/Exception.ooc]: This actually works. It will be included in the cleanup PR. Thanks !

**Comment by davidhesselbom** [comment]
The exception still needs freeing manually in the `catch` clause, though, right?

**Comment by sebastianbaginski** [comment]
Yes. Do you want me to make it automatic ?
Maybe in the upcoming *cleanup* PR as the code there is easier to work with.

**Comment by davidhesselbom** [comment]
Sure, it can wait, I was just asking.

**Comment by sebastianbaginski** [comment]
Sure, I think it is a good idea.

**Comment by davidhesselbom** [comment]
... but what happens if the `catch` throws the exception again?

**Comment by sebastianbaginski** [comment]
`raise` creates new `Exception` instance each time. Is there another way to throw exceptions, other than `raise` function ?

**Comment by davidhesselbom** [comment]
Well, you can create an exception manually and throw it, but the result should be the same.

**Comment by davidhesselbom** [comment]
Okay, just tested it. Leak is gone, finally, but I still get invalid reads.

**Comment by sebastianbaginski** [comment]
OK, thanks, I will try to address this issue in the next PR.

**Comment by marcusnaslund** [comment]
Woah, on Windows, `ExceptionTest` generates hundreds of lines of `[lang/Backtrace] No backtrace extension nor execinfo - use a debugger!` when running it.

**Comment by davidhesselbom** [comment]
Yeah, the library for backtraces is most likely missing on Windows.

**Comment by marcusnaslund** [comment]
It's fine - AppVeyor is still happy, it's just not a pretty output.

**Comment by davidhesselbom** [comment]
Well, either we put that printout inside a some kind of version block in order to ignore it, or we get the backtrace library working on Windows. The latter option will probably not be very easy, but I'm sure it's doable.

**Pull request merged**